### ISSUE TITLE
False ambiguity detection

### ISSUE THREAD

**Issue by timholy** [JuliaLang/julia#17350]
Adapted from Images.jl:
```jl
julia> using Colors

julia> immutable ScaleMinMax{To,From} end

julia> function map1{To<:Union{RGB24,ARGB32},From<:Real}(mapi::ScaleMinMax{To,From}, val::From)
           t = clamp(val, mapi.min, mapi.max)
           f = mapi.s*t - mapi.s*mapi.min
           convert(UFixed8, f)
       end
map1 (generic function with 1 method)

julia> function map1{To<:Union{RGB24,ARGB32},From<:Real}(mapi::ScaleMinMax{To,From}, val::Union{Real,Colorant})
           map1(mapi, convert(From, val))
       end
map1 (generic function with 2 methods)

julia> Base.Test.detect_ambiguities(Main)
1-element Array{Tuple{Method,Method},1}:
 (map1{To<:Union{ColorTypes.ARGB32,ColorTypes.RGB24},From<:Real}(mapi::ScaleMinMax{To,From}, val::From) at REPL[3]:2,map1{To<:Union{ColorTypes.ARGB32,ColorTypes.RGB24},From<:Real}(mapi::ScaleMinMax{To,From}, val::Union{ColorTypes.Colorant,Real}) at REPL[4]:2)
```
I think the first is strictly more specific than the second (even if "diagonality" is not considered).

**Comment by timholy** [comment]
Expanded and simplified (doesn't use `Colors`):
```jl
immutable ScaleMinMax{To,From} end

map1{To<:Union{Float32,Float64},From<:Real}(mapi::ScaleMinMax{To,From}, val::From) = 1
map1{To<:Union{Float32,Float64},From<:Real}(mapi::ScaleMinMax{To,From}, val::Union{Real,Complex}) = 2

map2{To<:Real,From<:Real}(mapi::ScaleMinMax{To,From}, val::From) = 1
map2{To<:Real,From<:Real}(mapi::ScaleMinMax{To,From}, val::Union{Real,Complex}) = 2

map3{To,From<:Real}(mapi::ScaleMinMax{To,From}, val::From) = 1
map3{To,From<:Real}(mapi::ScaleMinMax{To,From}, val::Union{Real,Complex}) = 2

map4{To<:Union{Float32,Float64},From<:Real}(mapi::ScaleMinMax{To,From}, val::Real) = 1
map4{To<:Union{Float32,Float64},From<:Real}(mapi::ScaleMinMax{To,From}, val::Union{Real,Complex}) = 2
```
Only `map1` is judged as ambiguous.

**Comment by timholy** [comment]
Git bisect identifies c779caf0ee823128d8410f8aaf3d2255231a6acb

**Comment by timholy** [comment]
@JeffBezanson, would it be worth also checking to see if your upcoming fix also solves #17354?
